sprintf strings

m4z opened the pull request:
This PHP noob is attempting to convert concat strings to sprintf, probably making it worse before it gets better…

Feel free to nudge (or 180) me into the right direction, or write superseding PRs. :wink:

Commit:
Attempt to get rid of string concatenation

Is this even an improvement?

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
This certainly feels worse

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
This one has already been converted

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented:
Addresses #6120.

m4z commented on Themes/default/languages/index.english.php:
```diff
@@ -150,7 +150,7 @@
 $txt['unnotify'] = 'Unnotify';
 // Use numeric entities in the below string.
-$txt['regards_team'] = "Regards,\nThe " . $context['forum_name'] . ' Team.';
+$txt['regards_team'] = 'Regards,' . "\n" . 'The %1$s Team.';
```
I don't understand enough about PHP (or the SMF requirements) to address the newline.

Oldiesmann replied:
Do the double quotes mess up the `%1$s` part? If not, `$txt['regards_team'] = "Regards,\nThe %1$s Team.";` should work.

m4z replied:
I thought they mess it up, but apparently [it should work like you suggest](https://www.php.net/manual/en/language.types.string.php#language.types.string.parsing).

m4z replied:
Hmm, but then again, this creates loads of `8: Undefined variable: s` errors in my test install… :cry:

m4z replied:
Just committed a potential fix (6fdc40e8f), which I'll test tomorrow…

Commit:
Less convoluted, as proposed by @Oldiesmann

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['welcome_guest_activate']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['template_parse_error_details']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented:
Ok, I have no clue how to do `$txt['movetopic_default']`; apparently the string is only loaded by `loadLanguage()` and then _maybe_ somehow used via `$_POST['reason']`… (not sure about that all all, though)

Commit:
sprintf()-ify $txt['*_topics_visit_none']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

jdarwood007 commented:
@m4z 
Its defined in this way
```
Zeus:Downloads jeremy$ grep -Rn "movetopic_default" ~/repos/sm/SMF2.1
~/repos/sm/SMF2.1/Sources/MoveTopic.php:111:		$temp = $txt['movetopic_default'];
~/repos/sm/SMF2.1/Sources/MoveTopic.php:114:		$txt['movetopic_default'] = $temp;
~/repos/sm/SMF2.1/Themes/default/languages/index.english.php:539:$txt['movetopic_default'] = 'This topic has been moved to ' . $txt['movetopic_auto_board'] . ".\n\n" . $txt['movetopic_auto_topic'];
```

And used in this way

```
							<dd>
								<textarea name="reason">', $txt[$type . 'topic_default'], '</textarea>
							</dd>
```

Which makes it easier to understand:
```
Zeus:Downloads jeremy$ grep -Rn "topic_default" ~/repos/sm/SMF2.1
/Users/jeremy/repos/sm/SMF2.1/Sources/MoveTopic.php:111:		$temp = $txt['movetopic_default'];
/Users/jeremy/repos/sm/SMF2.1/Sources/MoveTopic.php:114:		$txt['movetopic_default'] = $temp;
/Users/jeremy/repos/sm/SMF2.1/Themes/default/MoveTopic.template.php:103:								<textarea name="reason">', $txt[$type . 'topic_default'], '</textarea>
/Users/jeremy/repos/sm/SMF2.1/Themes/default/languages/index.english.php:539:$txt['movetopic_default'] = 'This topic has been moved to ' . $txt['movetopic_auto_board'] . ".\n\n" . $txt['movetopic_auto_topic'];
/Users/jeremy/repos/sm/SMF2.1/Themes/default/languages/index.english.php:695:$txt['mergetopic_default'] = 'This topic has been merged into ' . $txt['movetopic_auto_topic'] . '.';
```

So just convert both strings to be sprintf safe and this works.

Commit:
Attempt to obey Sensei @jdarwood007

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $forum_copyright

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
A little less concatenation, a little more action

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['email_variables']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented:
I can't find `$txt['attach_dir_files_missing']` being used anywhere.

Commit:
sprintf()-ify attachment related admin strings

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

jdarwood007 commented on Themes/default/MoveTopic.template.php:
```diff
@@ -99,8 +99,18 @@ function template_redirect_options($type)
+		echo '
+								<textarea name="reason">', sprintf($txt[$type . 'topic_default'], $txt['movetopic_auto_topic']), '</textarea>';
+	}
+	echo '
```
Since you specified in the language strings the numerical indexes (%1, %2), you can just use one call here.  The other replacement will be ignored.

Commit:
Simplify, as suggested by @jdarwood007

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Merge branch 'release-2.1' into sprintf-strings

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['xml_rss_desc']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['theme_thumbnail_href']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['who*'] in one fell swoop

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['forumProfile_info']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['forumProfile_info']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Don't concat $txt['profile_warning_notify_template_outline*']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['profile_warning_notify_template_outline*']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['profile_error_password_short']

I'm going with `if` here because while a single `sprintf` string would work
_now_, it would break the instant another `profile_error_*` string tries to
use a different argument.

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $helptxt['cal_short_*']

This is probably the most amateurish kludge imaginable, but at least it took
an hour to figure out…  ¯\_(ツ)_/¯

(I even tested this!)

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented on Themes/default/Help.template.php:
```diff
@@ -29,7 +29,16 @@ function template_popup()
-			', $context['help_text'], '<br>
+			';
+
+	if ($_GET['help'] == 'cal_short_months')
```
It might be more appropriate to do this at the bottom of `Sources/Help.php`, but I wasn't sure which of the three cases I needed and I feared I'd have to copy the `if` into each case. If I should move it there, tell me.

SychO9 replied:
yea It would definitely be much much better if you were to moves these conditional statements to the Source file.

The function you are looking for is `ShowAdminHelp()` located in `Help.php`. I haven't tested but I believe you want your code to be added before the last if statement in that function.

Instead of echoing you will just be assigning the result to `$context['help']`.

Also you might want to play with a `switch` block ?

MissAllSunday replied:
Indeed, a switch block would be preferable here.

Commit:
sprintf()-ify $helptxt['enableSpellChecking']

I think I made this worse.

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix syntax error

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['regards_team']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comments for $txt['welcome_guest*']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['template_parse_error_details']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comments for $txt['*_topics_visit_none']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $helptxt['pgFulltextSearch']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented on Themes/default/Help.template.php:
```diff
@@ -35,6 +35,8 @@ function template_popup()
 	elseif ($_GET['help'] == 'cal_short_days')
 		echo sprintf($context['help_text'], $txt['days_short'][1], $txt['days'][1]);
+	elseif ($_GET['help'] == 'enableSpellChecking')
+		echo sprintf($context['help_text'], ((function_exists('pspell_new') || function_exists('enchant_broker_init')) ? $helptxt['enableSpellCheckingSupported'] : $helptxt['enableSpellCheckingUnsupported']))
```
Is `$helptxt` even available in this context?

m4z commented on Themes/default/languages/Help.english.php:
```diff
@@ -195,8 +195,8 @@
-$helptxt['cal_short_months'] = 'If this setting is checked, month names within the calendar will be shortened.<br><br><strong>Enabled:</strong> ' . $txt['months_short'][1] . ' 1<br><strong>Disabled:</strong> ' . $txt['months_titles'][1] . ' 1';
-$helptxt['cal_short_days'] = 'If this setting is checked, day names within the calendar will be shortened.<br><br><strong>Enabled:</strong> ' . $txt['days_short'][1] . '<br><strong>Disabled:</strong> ' . $txt['days'][1];
+$helptxt['cal_short_months'] = 'If this setting is checked, month names within the calendar will be shortened.<br><br><strong>Enabled:</strong> %1$s 1<br><strong>Disabled:</strong> %2$s 1';
+$helptxt['cal_short_days'] = 'If this setting is checked, day names within the calendar will be shortened.<br><br><strong>Enabled:</strong> %1$s<br><strong>Disabled:</strong> %1$s';
```
Hello, Mr. Smartypants! "1" twice?

m4z replied:
Fixed by 005b66649

Commit:
`sprintf()`-ify `$helptxt['queryless_urls']`

Unsure if $helptxt is even available here, to use
`$helptxt['queryless_urls_not']`.

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix wrong placeholder in $helptxt['cal_short_days']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comments for $helptxt['cal_short_*']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $helptxt['who_enabled']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $helptxt['cron_is_real_cron']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['coppa_after_registration']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['coppa_form_title']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['coppa_form_body']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comments for $txt['*topic_default']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $forum_copyright

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['security_wrong']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['signature_settings_warning']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Err, good morning

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['modSettings_desc']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['edit_language_entries_desc']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['paid_note']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Single-string-ify

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix a false positive

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix #6128

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix #6132

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix 6125

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['backtrace_info']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Address the easy part of #6115

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify and un-duplicate $txt['ftp_path_info']

I have confirmed in a test install that the second one was used, however, it
doesn't seem to work; it opens a JS popup window with install.php?ftphelp,
but that doesn't do anything special, it's the same as install.php.

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Address the issue identified in 72cf0dd3900f843f0ecb86eec3db3bbe4dbd305d

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Remove context comment for $txt['ftp_path_info']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented:
This should be testable now; I have created a test install and will begin testing tomorrow.

Commit:
Fix syntax error

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix another syntax error in a double-quoted string

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix another syntax error in a double-quoted string

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix unused global variable

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix unused global variable

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix another syntax error in a double-quoted string

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix unused global variable

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix unused global variable

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Fix unused global variable

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

SychO9 commented on Themes/default/languages/Help.english.php:
```diff
@@ -381,7 +383,12 @@
 $helptxt['preview_characters'] = 'This setting sets the number of available characters for the first and last message topic preview.';
 $helptxt['quote_expand'] = 'This setting allows large quotes to be cutoff and expanded to their full length by the user. The setting indicates the minimum height a quote needs to have for it to be collapsed. Quotes smaller than this setting won\'t be affected';
 $helptxt['posts_require_captcha'] = 'This setting will force users to pass anti-spam bot verification each time they make a post to a board. Only users with a post count below the number set will need to enter the code - this should help combat automated spamming scripts.';
-$helptxt['enableSpellChecking'] = 'Enable spell checking. You MUST have the pspell or enchant library installed on your server and your PHP configuration set up to use the installed library. Your server ' . ((function_exists('pspell_new') || function_exists('enchant_broker_init')) ? 'DOES' : 'DOES NOT') . ' appear to have this set up.';
+// argument(s): either $helptxt['enableSpellCheckingSupported'] or $helptxt['enableSpellCheckingUnsupported'], depending on whether the server supports one of the libs
+$helptxt['enableSpellChecking'] = 'Enable spell checking. You MUST have the pspell or enchant library installed on your server and your PHP configuration set up to use the installed library. Your server %1$s appear to have this set up.';
+// is conditionally used as part of $helptxt['enableSpellChecking']
+$helptxt['enableSpellCheckingSupported'] = 'DOES';
+// is conditionally used as part of $helptxt['enableSpellChecking']
+$helptxt['enableSpellCheckingUnsupported'] = 'DOES NOT';
```
This bothers me a little xD

We really shouldn't be breaking phrases up like that, they won't be able to be properly translated this way.

even the way it was before the edit was wrong, what we should have is two text strings

One for when it is supported
```php
$helptxt['enableSpellCheckingSupported'] = 'Enable spell checking. You MUST have the pspell or enchant library installed on your server and your PHP configuration set up to use the installed library. Your server DOES appear to have this set up.';
```

The other for when it is not supported
```
$helptxt['enableSpellCheckingNotSupported'] = 'Enable spell checking. You MUST have the pspell or enchant library installed on your server and your PHP configuration set up to use the installed library. Your server DOES NOT appear to have this set up.';
```

It should not matter whether a big part of the string is practically duplicated, this is a better approach for the sake of sane translation, other languages translate very differently and thus imho we should not control the string with words this way.

SychO9 replied:
Or even better we should just separate the `Your server DOES/NOT appear to have this set up.` into two strings so the first bit won't be duplicated this way. and then we simply concatenate them in the source file.

m4z replied:
I attempted to move in that direction in f0afe70ad7cad3b7ce6638715f92f4aeaa9d4868, but I'm unsure how to change the [code](https://github.com/SimpleMachines/SMF2.1/pull/6121/commits/b75bac4b4c87bb30c2c2c5c1e1973e5281263bc9), since a single `$helptxt` variable has to be passed. Do you want it to be changed, so that instead of `enableSpellChecking`, `enableSpellChecking{Supported,Unsupported}` is being passed?

SychO9 replied:
The code actually works just fine with the change, `sprintf()` either injects `$helptxt['enableSpellCheckingSupported']` or `$helptxt['enableSpellCheckingUnsupported']` at the end of the text string which is exactly what we want.

unless I didn't understand what you meant ?

m4z replied:
What I meant was that `Sources/Help.php` is currently passing  "`enableSpellChecking`" to `Themes/default/Help.template.php`, and because I'm not 100% sure how to change that in `Help.php`, I still left it in the template until now. Will try to fix that today.

Commit:
Revert "Fix #6132"

This reverts commit 35ca7f70adab704fe5df2189a9df9f0fb061d318.

Commit:
Move towards what @SychO9 suggested in review

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

MissAllSunday commented on Themes/default/Profile.template.php:
```diff
@@ -2698,8 +2698,16 @@ function template_error_message()
 		// Cycle through each error and display an error message.
 		foreach ($context['post_errors'] as $error)
-			echo '
+			if ($error == 'password_short')
```
You can move the check for "password_short" to a ternary to avoid the if/else condition:

```php
		foreach ($context['post_errors'] as $error)
		{
			$text_key_error = $error == 'password_short' ?
				printf($txt['profile_error_' . $error], (empty($modSettings['password_strength']) ? 4 : 8)) :
				$txt['profile_error_' . $error];

				echo '
				<li>', isset($txt['profile_error_' . $error]) ? $text_key_error : $error, '</li>';
		}
```

m4z replied:
I find that hard to read, but you're the boss. (I think your `echo` was indented too far, changed that. (658f963))

Commit:
Attempt to address @MissAllSunday's review comment

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Move logic out of template

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
There's no endif in PHP

There's no endif in PHP
There's no endif in PHP
There's no endif in PHP

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Switch to switch

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['email_variables']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comments for $txt['basedirectory_for_attachments_warning'], $txt['attach_current_dir_warning']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['error_unexpected_template_call']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['theme_thumbnail_href']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Add context comment for $txt['xml_rss_desc']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Follow-up of fbf0b09114234485737217ad4bb92144b288df39

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
You can't actually use BBCode here

I guess this bug existed since e2e863284945e0dbce3e1d121616ae0a67af7eab.
Maybe this depends on some BBCode setting my test forum doesn't have, but it
_is_ the only BBCode in that file.

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
These are already handled by the following if

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Move both cases into their own sentences

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented:
I'm incapable to test `$txt['template_parse_error_details']`, I just can't manage to trigger it.

m4z commented:
I also don't know to create the conditions needed for `$txt['basedirectory_for_attachments_warning']` to show up; I attempted to empty `smf_settings:attachmentUploadDir`, but that just caused general errors.

Commit:
sprintf()-ify $txt['who_index']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
sprintf()-ify $txt['whoa*']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented:
I also have no clue how to test `$txt['backtrace_info']`.

Commit:
Add context comments for $helptxt['queryless_urls_*']

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

m4z commented:
* I have not yet tested each `who*` individually
* I'll have to do a tedious re-install to test the `switch`

Other than that (and my 3 previous comments), I have tested everything to the best of my abilities.

SychO9 commented on Sources/Help.php:
```diff
@@ -163,6 +163,24 @@ function ShowAdminHelp()
 	else
 		$context['help_text'] = $_GET['help'];
 
+	switch ($_GET['help']) {
+		case 'cal_short_months':
+			$context['help_text'] = printf($context['help_text'], $txt['months_short'][1], $txt['months_titles'][1]);
+			break;
+		case 'cal_short_days':
+			$context['help_text'] = printf($context['help_text'], $txt['days_short'][1], $txt['days'][1]);
+			break;
+		case 'cron_is_real_cron':
+			$context['help_text'] = printf($context['help_text'], $boarddir, $boardurl);
+			break;
+		case 'enableSpellChecking':
+			$context['help_text'] = printf($context['help_text'], ((function_exists('pspell_new') || function_exists('enchant_broker_init')) ? $helptxt['enableSpellCheckingSupported'] : $helptxt['enableSpellCheckingUnsupported']));
+			break;
+		case 'queryless_urls':
+			$context['help_text'] = printf($context['help_text'], (isset($_SERVER['SERVER_SOFTWARE']) && (strpos($_SERVER['SERVER_SOFTWARE'], 'Apache') !== false || strpos($_SERVER['SERVER_SOFTWARE'], 'lighttpd') !== false) ? $helptxt['queryless_urls_supported'] : $helptxt['queryless_urls_unsupported']));
+			break;
+	}
```
please change `printf` to `sprintf` you don't want to print them right there and then, you just want to replace the proper bits of the strings

m4z replied:
I think I did that because of 658f963ba11b0935557764e08befe2fc0e84a369, should I change it there as well?

m4z replied:
Fixed in 7b61f1f

SychO9 replied:
Thanks, and no in the template files it's fine since printing there is intended, but we (normally) don't print anything from the source files. 👍

SychO9 commented on Themes/default/languages/Who.english.php:
```diff
@@ -117,24 +141,36 @@
 $txt['whoallow_search2'] = 'Viewing the results of a search.';
-$txt['whoallow_stats'] = 'Viewing the <a href="' . $scripturl . '?action=stats">forum stats</a>.';
+// argument(s): $scripturl
+$txt['whoallow_stats'] = 'Viewing the <a href="%1$s?action=stats">forum stats</a>.';
```
none of the `whoallow_` ones are being replaced, I believe you need to do so here

https://github.com/SimpleMachines/SMF2.1/blob/017e10b91144d1d68754947e6364944c9902b348/Sources/Who.php#L428

Commit:
Address review by @SychO9

https://github.com/SimpleMachines/SMF2.1/pull/6121#discussion_r433916089

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

Commit:
Address review by @SychO9

https://github.com/SimpleMachines/SMF2.1/pull/6121#discussion_r433983719

Signed-off-by: Christopher 'm4z' Holm <them4z@gmail.com>

SychO9 reviewed (approved):
Apart from the two things I noticed in my previous review, everything else looks good to me, I did my best and tested every single edited text string.

Nicely done @m4z

MissAllSunday commented:
Is this considered done?

m4z commented:
> Apart from the two things I noticed in my previous review, everything else looks good to me, I did my best and tested every single edited text string.

Awesome, thanks!


> Is this considered done?

I think so. I didn't get to testing the rest of the changes yet, but IIUC, @SychO9 did it for me.

halasla reviewed (commented):
--SPAM, But Github doesn't let you delete spam.

m4z commented:
Have you reported the account?
